### PR TITLE
ENG-10962 graceful shutdown

### PR DESCRIPTION
**Pull request by zeemanhe**
--refuse new transactions during shutdown
--drain export and dr buffer
--waiting for all outstanding transactions to be completed.

### CODE REVIEW

File: lib/python/voltcli/checkstats.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+     while True:
+        resp = get_stats(runner, 'LIVECLIENTS')
+        trans = 0
+        bytes = 0
+        msgs  = 0
+        for r in resp.table(0).tuples():
+            bytes += r[6]
+            msgs += r[7]
+            trans += r[8]
+        runner.info('Outstanding transactions=' + str(trans) + ', buffer size=' + str(bytes) + ', response messages=' + str(msgs))
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Maybe worth repeating the "Outstanding" for "request bytes", "response messages"? Following the keywords on 'LIVECLIENTS'.

File: lib/python/voltcli/checkstats.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+def get_stats(runner, component):
+    retry = 5
+    while True:
+        response = runner.call_proc('@Statistics', [voltdbclient.FastSerializer.VOLTTYPE_STRING,
+                                    voltdbclient.FastSerializer.VOLTTYPE_INTEGER], [component, 0])
+        status = response.status()
+        if status <> 1 and "timeout" in response.statusString:
+            if retry == 0:
+                runner.error('Unable to collect statistics from the cluster')
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Maybe put the "component" into the error message as before?

File: src/frontend/org/voltdb/InvocationDispatcher.java [hunk 3]
```diff
@@ -501,6 +503,11 @@ else if ("@UpdateClasses".equals(procName)) {
     }
 
     private final static boolean allowPauseModeExecution(InvocationClientHandler handler, Procedure procedure, StoredProcedureInvocation invocation) {
+        if (VoltDB.instance().getMode() == OperationMode.PAUSED && VoltDB.instance().isShuttingdown()
+                && !procedure.getAllowedinshutdown()) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/InvocationDispatcher.java]: Can you add a comment to explain what are procedures should be allowed during pause & shutdown mode?

File: src/frontend/org/voltdb/InvocationDispatcher.java [hunk 4]
```diff
@@ -501,6 +503,11 @@ else if ("@UpdateClasses".equals(procName)) {
     }
 
     private final static boolean allowPauseModeExecution(InvocationClientHandler handler, Procedure procedure, StoredProcedureInvocation invocation) {
+        if (VoltDB.instance().getMode() == OperationMode.PAUSED && VoltDB.instance().isShuttingdown()
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/InvocationDispatcher.java]: If you move this IF check after line 511, you do not need to check the instance mode again.
Like
if (VoltDB.instance().isShuttingdown() && && !procedure.getAllowedinshutdown()) {
    return false;
}

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 5]
```diff
@@ -315,6 +316,15 @@ public boolean isRunningWithOldVerbs() {
         return m_isRunningWithOldVerb;
      };
 
+     @Override
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/RealVoltDB.java]: Weird indentation.

File: tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        resp = client.callProcedure("@Statistics", "liveclients", 0);
+        assertTrue(resp.getStatus() == ClientResponse.SUCCESS);
+
+        try {
+            client.callProcedure("@SystemInformation", "OVERVIEW");
+            fail();
+        } catch (ProcCallException e) {
+        }
+    }
+
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Can you add a long running query 10 seconds maybe, then you can test the graceful shutdown with waiting in-flight transactions? Also add tests for rejecting new transactions.

These two are easily unit testable. Not sure about DR and export. I think we have system tests for them, which wiebe can focus on.

Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Another test case can be: issue other system procedure that is not allowable during shutdown process.

**Comment by akhanzode** [comment]
You need to wait for importer queued transactions to complete before you can say shutting down to true otherwise those transactions will be lost. You can pause and then monitor importer stats pending to go down to 0. Are all read only transactions allowed? This passthrough looks like PermissionValidator thingy you should hide allow of execution in there.

File: tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java (hunk 6 above)
Comment (zeemanhe) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Test case updated.

**Comment by zeemanhe** [comment]
Importer stats is now checked.  Read-only transactions but @Statistics are currently not allowed in non-admin mode.  Should Read-only transactions be blocked for admin too?

File: lib/python/voltcli/checkstats.py [hunk 7]
```diff
@@ -185,7 +186,18 @@ def check_clients(runner):
             bytes += r[6]
             msgs += r[7]
             trans += r[8]
-        runner.info('Outstanding transactions=' + str(trans) + ', buffer size=' + str(bytes) + ', response messages=' + str(msgs))
+        runner.info('Outstanding transactions=%d, Outstanding buffer bytes=%d, Outstanding response messages=%d' %(trans, bytes,msgs))
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: REQUEST_BYTES

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 8]
```diff
@@ -13,11 +13,35 @@
...
+    )
 )
 def shutdown(runner):
+    if runner.opts.forcing==False:
+        runner.info('Preparing for shutdown')
+        status = runner.call_proc('@Pause', [], []).table(0).tuple(0).column_integer(0)
+        if status <> 0:
+            runner.error('The preparation for shutdown failed with status: %d' % status)
+            return
+        runner.info('The cluster has been paused and will not accept new transactions.')
```
Comment (xinjiacs) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: "new write transactions" maybe?

**Comment by xinjiacs** [comment]
Looks good to me.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+{
+    @Override
+    public void init() {}
+
+    @Override
+    public DependencyPair executePlanFragment(
+            Map<Integer, List<VoltTable>> dependencies, long fragmentId,
+            ParameterSet params, SystemProcedureExecutionContext context)
+    {
+        throw new RuntimeException("PrepareShutdown was given an " +
```
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: IllegalStateException?

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+                throw new RuntimeException(e);
+            }
+            VoltDB.instance().setShuttingdown(true);
+        }
+
+        VoltTable t = new VoltTable(VoltSystemProcedure.STATUS_SCHEMA);
+        t.addRow(VoltSystemProcedure.STATUS_OK);
+        return (new VoltTable[] {t});
+    }
+}
```
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: Need a newline.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import org.voltdb.ParameterSet;
+import org.voltdb.ProcInfo;
+import org.voltdb.SystemProcedureExecutionContext;
+import org.voltdb.VoltDB;
+import org.voltdb.VoltSystemProcedure;
+import org.voltdb.VoltTable;
+import org.voltdb.VoltZK;
+
+@ProcInfo(singlePartition = false)
+public class PrepareShutdown extends VoltSystemProcedure
```
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: Should this whole class inherit from Pause? The code seems different by about 2 lines.

File: tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+import org.voltdb.client.ProcedureCallback;
+import org.voltdb.compiler.VoltProjectBuilder;
+
+public class TestPrepareShutdown extends RegressionSuite
+{
+    public TestPrepareShutdown(String name) {
+        super(name);
+    }
+
+    public void testPrepareShutdown() throws Exception {
```
Comment (jhugg) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Testing things I'd like to see:
1. Fill export buffers, safe-shutdown, verify export data is in downstream.
2. Run DR2 second cluster. Flood first cluster with writes. safe-shutdown, verify DR data is downstream.
3. Flood with slow writes. prepare-shutdown. Kill a node before actual shutdown. Verify you can recover.

Comment (jhugg) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Might want to see if eco team already has some similar tests that could be modified to work here? (crossing-fingers)

Comment (zeemanhe) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: Test cases have been added to /pro with export, dr and recovery

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 13]
```diff
@@ -13,11 +13,37 @@
...
+    )
 )
 def shutdown(runner):
+    if runner.opts.forcing==False:
+        runner.info('Preparing for shutdown')
+        status = runner.call_proc('@PrepareShutdown', [], []).table(0).tuple(0).column_integer(0)
+        if status <> 0:
+            runner.error('The preparation for shutdown failed with status: %d' % status)
+            return
+        runner.info('The cluster has been paused and will not accept new write transactions.')
```
Comment (zeemanhe) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: Stats ware allowed. Other reads will be blocked.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 11 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: make sense. PrepareShutdown extends Pause now.

File: src/frontend/org/voltdb/VoltDBInterface.java [hunk 14]
```diff
@@ -87,7 +87,8 @@
      * @return
      */
     public boolean shutdown(Thread mainSiteThread) throws InterruptedException;
-
+    public boolean isShuttingdown();
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/VoltDBInterface.java]: comments were added

File: tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+            }
+            System.out.printf("Outstanding importer transactions: %d\n", sum);
+            Thread.sleep(500);
+        }
+        assertTrue (sum == 0);
+        try{
+            client.callProcedure("@Shutdown");
+            fail("@Shutdown fails via admin mode");
+        } catch (ProcCallException e) {
+            //if execution reaches here, it indicates the expected exception was thrown.
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: @zeemanhe Can you also assert the error message here?

Comment (zeemanhe) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: done

File: lib/python/voltcli/voltadmin.d/shutdown.py (hunk 13 above)
Comment (jhugg) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: Shouldn't the error message say this then?

Comment (zeemanhe) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: message changed to 'The cluster has been paused for shutdown and will not accept new transactions other than @Statistics'

Comment (jhugg) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: Ok, last dumb request here. Can you run this by Andrew Gent?

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 11 above)
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: Still waiting on a comment here.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 9 above)
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: Ping.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 10 above)
Comment (jhugg) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: Ping.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 11 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: The comments were added but not displayed here.  you may view the file to checkout the comments.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 10 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: The file was updated. view the file to checkout the change.

File: src/frontend/org/voltdb/sysprocs/PrepareShutdown.java (hunk 9 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/sysprocs/PrepareShutdown.java]: PrepareShutdown extends Pause which handles the exception.

**Comment by jhugg** [comment]
I'm good if Andrew signs off on new error messages.

**Comment by zeemanhe** [comment]
The message was updated per Andrew's suggestion:  
"The cluster is paused prior to shutdown and will not accept new transactions other than @Statistics."

**Pull request merged**